Alerting: Resend resolved notifications for ResolvedRetention duration

JacobsonMT opened the pull request:
**What is this feature?**

This feature introduces a new configuration option: `[unified_alerting].resolved_alert_retention=15m`. This configuration replicates the functionality of the Prometheus Ruler, where resolved notifications are sent to the Alertmanager multiple times over a span of 15 minutes (refer to the Prometheus source code [here](https://github.com/prometheus/prometheus/blob/1dd0bff4f17ce95c78d9aad5658c837573dbb5dd/rules/alerting.go#L454-L467)).

Key changes to alert instance state fields (in memory, not persisted):

- Removed: `Resolved bool`
  - Previously, this field was set to `true` during the state transition that resolved a firing alert. It was only `true` for the exact state transition and `false` afterwards.
- Added: `ResolvedAt time.Time`
  - This field is set when the state first resolves, i.e., transitions from `Alerting`, `NoData`, or `Error` to `Normal`. It is reset to zero when the state transitions from `Normal` to any other state. This means the value is nonzero only when we have a `Normal` state that has transitioned from a firing state. `Normal` states warmed from the database or newly added to the cache will have `ResolveAt=zero`.

**Why do we need this feature?**

This feature enhances the resilience of resolved notifications against network issues. Currently, if the initial resolved notification fails to reach the Alertmanager, it is never resent, leading to an inconsistent state between Grafana and the Alertmanager. This results in the firing alert remaining orphaned in the Alertmanager until it naturally expires.

Additionally, this change paves the way for improved resilience against Grafana crashes or restarts that occur between when the state is persisted to the database and when the resolved notification is sent to the Alertmanager. In such cases, Grafana and the Alertmanager would also be left in an inconsistent state. This improvement will be addressed in a future update.

**Who is this feature for?**

Users of Grafana Alerting.

**Fixes**

Also a proposed mitigation for: https://github.com/grafana/grafana/issues/80996

**Special notes for your reviewer:**

- This feature may significantly increase the number of alerts sent to Alertmanagers. While this may not be a major issue for internal Alertmanagers, it could be notable for external or remote Alertmanagers.
- The feature can be disabled by setting `[unified_alerting].resolved_alert_retention=0` in the configuration.

Please check that:
- [x] It works as expected from a user's perspective.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

JacobsonMT committed:
Retain ResolvedAt time between Normal->Normal transition

JacobsonMT committed:
Introduce ResolvedRetention to keep sending recently resolved alerts

JacobsonMT committed:
Simple replace of State.Resolved with State.ResolvedAt

JacobsonMT committed:
Make ResolvedRetention configurable with resolved_alert_retention

JacobsonMT committed:
Tick-based LastSentAt for testing of ResendDelay and ResolvedRetention

JacobsonMT committed:
Cleanup

JacobsonMT commented on pkg/services/ngalert/state/compat.go:
```diff
@@ -139,13 +140,13 @@ func errorAlert(labels, annotations data.Labels, alertState *State, urlStr strin
 	alerts := apimodels.PostableAlerts{PostableAlerts: make([]models.PostableAlert, 0, len(firingStates))}
 	var sentAlerts []*State
-	ts := time.Now()
+	ts := evaluatedAt
```
NOTE: This change is not strictly necessary, though I believe it is more correct than before and is useful for creating tests for this change, as other times use a mocked clock.

JacobsonMT commented on pkg/services/ngalert/state/manager_private_test.go:
```diff
@@ -1304,6 +1304,7 @@ func TestProcessEvalResults_StateTransitions(t *testing.T) {
 									StartsAt:           t2,
 									EndsAt:             t2,
 									LastEvaluationTime: t3,
+									ResolvedAt:         t2,
```
NOTE: this shows retaining the `ResolvedAt` time through a normal->normal transition.

JacobsonMT commented on pkg/services/ngalert/state/manager_private_test.go:
```diff
@@ -2330,6 +2331,7 @@ func TestProcessEvalResults_StateTransitions(t *testing.T) {
 									StartsAt:           t2,
 									EndsAt:             t2,
 									LastEvaluationTime: t3,
+									ResolvedAt:         t2,
```
NOTE: this shows retaining the `ResolvedAt` time through a normal->normal transition.

JacobsonMT commented on pkg/services/ngalert/state/state.go:
```diff
@@ -400,19 +395,38 @@ func resultKeepLast(state *State, rule *models.AlertRule, result eval.Result, lo
+// - The state has been resolved since the last notification.
+// - The state is firing and the last notification was sent at least resendDelay ago.
+// - The state was resolved within the resolvedRetention period, and the last notification was sent at least resendDelay ago.
+func (a *State) NeedsSending(resendDelay time.Duration, resolvedRetention time.Duration) bool {
```
NOTE: Decent amount of logic changes here, but at the very least it brings us closer to the logic in prometheus ruler.

yuri-tceretian replied:
Yes, but in my opinion it is harder to read now.

JacobsonMT commented on pkg/services/ngalert/schedule/alert_rule_test.go:
```diff
@@ -762,8 +763,95 @@ func TestRuleRoutine(t *testing.T) {
 		require.NotEmpty(t, sch.stateManager.GetStatesForRuleUID(rule.OrgID, rule.UID))
 	})
+
+	t.Run("when there are resolved alerts they should keep sending until retention period is over", func(t *testing.T) {
```
NOTE: Main test showing multiple sends for a resolved notification and how it interacts with resend delay.

yuri-tceretian commented on pkg/services/ngalert/state/manager.go:
```diff
@@ -247,7 +252,11 @@ func (st *Manager) DeleteStateByRuleUID(ctx context.Context, ruleKey ngModels.Al
 		// to Alertmanager.
-		s.Resolved = oldState == eval.Alerting || oldState == eval.Error || oldState == eval.NoData
+		if oldState == eval.Alerting || oldState == eval.Error || oldState == eval.NoData {
+			s.ResolvedAt = now
```
What about `pending`? Let's imagine a situation when:
`Normal -> Pending -> Alerting -> Normal -> Pending -> Normal`

In the case of the second and thrid `Normal` we won't send the resolved notification.

JacobsonMT replied:
The 2nd Normal will always send because we now check if `a.ResolvedAt.After(a.LastSentAt)` in `NeedsSending(...)`.

The 3rd Normal I'm on the fence about. Currently, I reset `ResolvedAt` when entering Pending as this follows how Prom Ruler does this. However, this isn't exactly obvious IMO and had to look it up to see what they decided to do here.

I lean slightly towards not resetting `ResolvedAt` when entering Pending, but I'd have to give some through to situations where this might have unintended consequences.

yuri-tceretian commented on pkg/services/ngalert/schedule/alert_rule_test.go:
```diff
@@ -762,8 +763,95 @@ func TestRuleRoutine(t *testing.T) {
+	il := models.InstanceLabels(s.Labels)
+	s.Labels = data.Labels(il)
+	id, err := il.StringKey()
+	require.NoError(t, err)
```
This has just been changed in another PR. Let's update to use labels' fingerprint

JacobsonMT committed:
More state transition tests

JacobsonMT committed:
Do not reset ResolvedAt during Normal->Pending transition

Initially this was done to be inline with Prom ruler. However, Prom ruler
doesn't keep track of Inactive->Pending/Alerting using the same alert instance,
so it's more understandable that they choose not to retain ResolvedAt. In our
case, since we use the same cached instance to represent the transition, it
makes more sense to retain it.

This should help alleviate some odd situations where temporarily entering
Pending will stop future resolved notifications that would have happened
because of ResolvedRetention.

JacobsonMT committed:
Fix merge conflicts

yuri-tceretian commented on pkg/services/ngalert/state/state.go:
```diff
@@ -400,19 +395,38 @@ func resultKeepLast(state *State, rule *models.AlertRule, result eval.Result, lo
+// RecentlyResolved returns true if the state has been resolved within the resolvedRetention period. Will return false
+// if the state is not resolved or if the state was resolved outside the resolvedRetention period.
+func (a *State) RecentlyResolved(resolvedRetention time.Duration) bool {
+	return !a.ResolvedAt.IsZero() && a.LastEvaluationTime.Sub(a.ResolvedAt) <= resolvedRetention
```
I think this can be a problem if rule evaluation interval is greater than resolvedRetention period. 
On other hand, I am not sure that there is a benefit of resending after such a long interval.

yuri-tceretian replied:
This is covered by check in NeedsResending. WDYT about moving it here just to keep all logic related to resending resolved alert in one place?

JacobsonMT replied:
Sure, I can inline it. It was easier to grok while developing like this, but now that it's fully commented it feels unnecessary.

yuri-tceretian reviewed (commented):
The solution seems to be good for rule which evaluation interval is relatively short. However, I do not think it works for rules with longer evaluation intervals, like 30 minutes or 1 hour, which are quite common.

I think that to properly solve it we need state manager to work asynchronously to the scheduler, and tick independently. This way we will not be bound to rule evaluation interval. 

However, I do not want to block this PR. I would like to do some testing and then I will approve it.

yuri-tceretian commented on pkg/services/ngalert/state/state.go:
```diff
@@ -400,19 +395,38 @@ func resultKeepLast(state *State, rule *models.AlertRule, result eval.Result, lo
+	}
+
+	// We should send a notification if the state has been resolved since the last notification.
+	if a.ResolvedAt.After(a.LastSentAt) {
```
ok, this is supposed to cover the case when resolve notification was never sent and the interval expired.
There are two problems.
1. LastSentAt is not saved to db right now.  PR https://github.com/grafana/grafana/pull/89135 addresses it but does not address the situation fully because LastSentAt is updated _after_ instances are saved into database. (this is the most annoying part of the state management, IMO)
2. It is not limited. I.e. if instance starts in a couple of days, this will cause the alert to be created. So, effectively, we do not honor resolved retention.

JacobsonMT replied:
1. Agreed, the fact that we update LastSentAt after saving the state is awkward. If we're going to address this (and I think we should) it should be done in the [persistence PR](https://github.com/grafana/grafana/pull/89135)
2. Just to be clear, this should never create a `Firing` alert as only `Normal` states can have a non-zero `ResolvedAt`. That being said, it will send an expired alert to alertmanager, as intended with this change, as the associated `Firing` alert in the AM was never resolved. The only optimization here could be to check if the previous `Firing` alert has already expired, but we don't store the necessary information to do this from a warmed cache.

JacobsonMT committed:
Inline state.RecentlyResolved(...)

JacobsonMT commented on pkg/services/ngalert/state/manager.go:
```diff
@@ -245,7 +250,11 @@ func (st *Manager) DeleteStateByRuleUID(ctx context.Context, ruleKey ngModels.Al
+		if oldState == eval.Alerting || oldState == eval.Error || oldState == eval.NoData {
+			s.ResolvedAt = now
+		} else {
+			s.ResolvedAt = time.Time{}
```
I mostly decided on `time.Time{}` to both stay inline with `LastSentAt` which doesn't use a pointer and to keep closer to Prom ruler as it creates a much cleaner `NeedsSending` logic. For example, if both  `LastSentAt` and `ResolvedAt` were pointers, we'd need:

```go
	// We should send a notification if the state has been resolved since the last notification.
	if a.ResolvedAt != nil && (a.LastSentAt == nil || a.ResolvedAt.After(*a.LastSentAt)) {
		return true
	}
```
 instead of:

```go
	// We should send a notification if the state has been resolved since the last notification.
	if a.ResolvedAt.After(a.LastSentAt) {
		return true
	}
```
If I just use a pointer to `ResolvedAt` it's still fairly clean, however then it makes persisting them awkward as they'll need different treatment. WDYT?

yuri-tceretian reviewed (approved):
LGTM. Great improvement.

JacobsonMT committed:
Pointers for ResolvedAt & LastSentAt

To avoid awkward time.Time{}.Unix() defaults on persist

yuri-tceretian reviewed (approved):
Still LGTM 🚢
